xdsl/transforms/csl_wrapper_to_csl.py

```diff
@@ -0,0 +1,144 @@
+
+        assert isa(yield_op := op.layout_module.block.last_op, csl_wrapper.YieldOp)
+        yield_op.detach()
+
+        with ImplicitBuilder(module_block):
+            const_0 = arith.Constant.from_int_and_width(0, builtin.IndexType())
+            const_1 = arith.Constant.from_int_and_width(1, builtin.IndexType())
+
+            const_width = arith.Constant(op.width)
+            param_width = csl.ParamOp("width", op.width.type, const_width)
+
+            const_height = arith.Constant(op.height)
+            param_height = csl.ParamOp("height", op.height.type, const_height)
+
+            for param in op.params:
+                if isa(param.value, builtin.IntegerAttr):
+                    value = arith.Constant(param.value)
+                else:
+                    value = None
+                p = csl.ParamOp(param.key.data, param.type, value)
+                new_args.append(SSAValue.get(p))
+
+            layout = csl.LayoutOp(Region())
+            with ImplicitBuilder(layout.body.block):
+                scf.For(
+                    lb=const_0,
+                    ub=const_width,
+                    step=const_1,
+                    iter_args=[x],
+                    body=outer_loop_block,
+                )
+
+                with ImplicitBuilder(outer_loop_block):
+                    scf.For(
+                        lb=const_0,
+                        ub=const_height,
+                        step=const_1,
+                        iter_args=[y],
+                        body=inner_loop_block,
+                    )
+        rewriter.inline_block(
+            # TODO(dk949): the clone is temporary
+            op.layout_module.block,
+            # op.layout_module.clone().block,
+            InsertPoint.at_start(inner_loop_block),
+            arg_values=[
+                SSAValue.get(param_width),
+                SSAValue.get(param_height),
+                SSAValue.get(x),
+                SSAValue.get(y),
+                *new_args,
+            ],
+        )
+        tile_code = csl.SetTileCodeOp(
+            fname=self.prog_name,
+            x_coord=x,
+            y_coord=y,
+        )
+        inner_loop_block.add_op(tile_code)
+
+        layout_mod = csl.CslModuleOp(
+            regions=[Region(module_block)],
+            properties={"kind": csl.ModuleKindAttr(csl.ModuleKind.LAYOUT)},
+        )
+        print(layout_mod)
+        exit(0)
+
+
+@dataclass(frozen=True)
+class CslWrapperToCslPass(ModulePass):
+    """
+    Wraps program in the csl_stencil dialect in a csl_wrapper by translating each
+    top-level function to one module wrapper.
+    """
+
+    name = "csl-wrapper-to-csl"
+    prog_name: str
+
+    def apply(self, ctx: MLContext, op: builtin.ModuleOp) -> None:
+        module_pass = PatternRewriteWalker(
+            GreedyRewritePatternApplier(
+                [
+                    RemoveUnusedYield(),
+                    ExtractLayoutProgramModule(prog_name=self.prog_name),
```

Comment:
You could read this from `csl_wrapper.module.program_name`

Reply:
:+1: 

I did keep the `prog_name` in as a back up, since `op.program_name` can be `None`, but I'm very much not marries to it

Reply:
Given we have it already, my preference would be not to have it as a command line arg unless there's genuine need for it.